### PR TITLE
feat: Prevention - On the fly quality checks in the product edit form

### LINKED ISSUE #1905: Prevention - On the fly quality checks in the product edit form (javascript)
### What
Create on the fly quality check for the web

- [ ] Saturated fat > Total fat
- [ ] Sugar > Glucids

### Part of 
- #5538

### PR DESCRIPTION
**Pull request by jnsereko**
### What
Created an invisible span having a warning on saturated fats and sugars values.  inputing an element above the range displays the span.

### Screenshot

https://user-images.githubusercontent.com/58003327/234127304-bb7f3a63-b37d-43d3-a8ca-8a758112b6ac.mov

### Related issue(s) and discussion

- Fixes #1905

### CODE REVIEW

File: templates/web/pages/product_edit/product_edit_form_display.tt.js [hunk 1]
```diff
@@ -34,3 +34,21 @@
 		\$('#nutrition_data_div').show();
 	}
 });
+
+  \$('#nutriment_sugars').on('input', function() {
+    var inputValue = $(this).val();
```
Comment (jnsereko) [inline templates/web/pages/product_edit/product_edit_form_display.tt.js:39]: hey @stephanegigandet @yuktea 
I jQuery is blocking this from working 

its like the `.val()` is not a function.   `Uncaught TypeError: this.val is not a function`

think i have called the element correctly so it must be working

Comment (stephanegigandet) [reply templates/web/pages/product_edit/product_edit_form_display.tt.js:39]: Hi @jnsereko , it's because this file is a template file, so anything starting with $ is interpreted as a variable in the template. If you want the $ to be output in the JS code, you need to escape it like this: \$

Comment (jnsereko) [reply templates/web/pages/product_edit/product_edit_form_display.tt.js]: thanks @stephanegigandet

**Comment by stephanegigandet** [comment]
> this @stephanegigandet somewhat messes up the table. i thought of creating a warning div, just like https://github.com/jnsereko/openfoodfacts-server/blob/d9631fdd0e73ac311a9512ebf156533369f775ae/templates/web/pages/product_edit/product_edit_form_display.tt.html#L78-L82 but since the table is somewhat long, covering almost the whole screen users might not see it

One option could be to have an extra table row with the warning, just below the value that is problematic. This way you could use the whole row width (the 3 columns).

**Comment by stephanegigandet** [comment]
Hi @jnsereko , this looks quite good, in terms of functionality and how to display the warning, it's great I think.

File: templates/web/pages/product_edit/product_edit_form_display.tt.js [hunk 2]
```diff
@@ -34,3 +34,66 @@
...
+		\$('#nutriment_question_mark_'+nutirent_id).css("display", "none");
+	}
+}
+
+var sugars_value;
+var carbohydrates_value;
+var saturated_fats_value;
+var fat_value;
+
+[% FOREACH nutriment IN nutriments %]
```
Comment (stephanegigandet) [inline templates/web/pages/product_edit/product_edit_form_display.tt.js:54]: This works, but it will generate a huge JS file, with the code repeated for all nutrients.

It would be better to create a javascript array that contains all nutrients, and then have a for loop in Javascript, instead of a foreach loop in the template language.

**Comment by CharlesNepote** [comment]
Well done @jnsereko! The video looks really cool. Last point: I would choose a lighter color to highlight errors, not to frighten people. The yellow is a bit flashy...

@stephanegigandet would it be possible to test it on openfoodfacts.net?

**Comment by jnsereko** [comment]
> Well done @jnsereko! The video looks really cool. Last point: I would choose a lighter color to highlight errors, not to frighten people. The yellow is a bit flashy...
> 
> @stephanegigandet would it be possible to test it on openfoodfacts.net?

hey @CharlesNepote 
Would pale pink work better?
Screenshot 2023-04-27 at 02 39 34

File: templates/web/pages/product_edit/product_edit_form_display.tt.js [hunk 3]
```diff
@@ -34,3 +34,67 @@
...
+}
+
+var sugars_value;
+var carbohydrates_value;
+var saturated_fats_value;
+var fat_value;
+
+var required_nutrients_id = ['energy-kj', 'energy-kcal', 'fat', 'saturated-fat', 'sugars', 'carbohydrates', 'fiber', 'proteins', 'salt', 'sodium', 'alcohol'];
+
+required_nutrients_id.forEach(nutirent_id => {
```
Comment (jnsereko) [inline templates/web/pages/product_edit/product_edit_form_display.tt.js:56]: @stephanegigandet, using the ordinary for loop is tricky.  Only the last nutrient, "alcohol" is accessed the colour but the `on-input` event exists on other elements 

Is this okay?

Comment (stephanegigandet) [reply templates/web/pages/product_edit/product_edit_form_display.tt.js:56]: It looks good this way, thank you! :)

Could you just replace all the "nutirent" by "nutrient" ? :)

Comment (jnsereko) [reply templates/web/pages/product_edit/product_edit_form_display.tt.js]: thanks @stephanegigandet 
That slipped my eyes

**Comment by jnsereko** [comment]
hey @stephanegigandet @alexgarel 
All tests are passing now 😄

**Review by stephanegigandet — APPROVED** [review]

**Comment by stephanegigandet** [comment]
@jnsereko Thank you!

File: templates/web/pages/product_edit/product_edit_form_display.tt.js [hunk 4]
```diff
@@ -34,3 +34,67 @@
...
+var saturated_fats_value;
+var fat_value;
+
+var required_nutrients_id = ['energy-kj', 'energy-kcal', 'fat', 'saturated-fat', 'sugars', 'carbohydrates', 'fiber', 'proteins', 'salt', 'sodium', 'alcohol'];
+
+required_nutrients_id.forEach(nutrient_id => {
+	\$('#nutriment_' + nutrient_id).on('input', function() {
+		var nutrient_value = \$(this).val();
+		var is_above_or_below_100 = isNaN(nutrient_value) || nutrient_value < 0 || nutrient_value > 100;
+		show_warning(is_above_or_below_100, nutrient_id, "Please enter a value between 0 and 100");
```
Comment (stephanegigandet) [inline templates/web/pages/product_edit/product_edit_form_display.tt.js:60]: The warnings should be added to po/common/common.pot and po/common/en.po so they can be translated, but we can also do that later in another PR.

File: tests/integration/fix_non_normalized_codes.t [hunk 5]
```diff
@@ -104,10 +104,10 @@ is_deeply(
 	[
 		# removed product_broken_code
 		"Removed broken-123",
-		# removed product with existing id
-		"Removed 0000012345670 as duplicate of 12345670",
 		# product_non_normalized_code moved
 		"Moved 0000012345678 to 12345678",
+		# removed product with existing id
```
Comment (stephanegigandet) [inline tests/integration/fix_non_normalized_codes.t:109]: This is a bug that is not related to your PR, could you remove this change?
The bug is fixed in this PR: https://github.com/openfoodfacts/openfoodfacts-server/pull/8384

File: templates/web/pages/product_edit/product_edit_form_display.tt.js (hunk 4 above)
Comment (jnsereko) [reply templates/web/pages/product_edit/product_edit_form_display.tt.js:60]: @stephanegigandet can we do the translations in another PR
i have tried them locally but it might take me some time to make the logic work.

**Comment by CharlesNepote** [comment]
The question mark should not appear when there is a "hyphen", meaning this data is not displayed on the packaging. It's important.
Should I open another bug or not?

See:
image
